### PR TITLE
Fix bug when combining 2+ mirror symmetries and Bloch-periodic boundary conditions

### LINKED ISSUE #132: bug when combining 3 mirror symmetries with periodic boundary conditions in 3d
As has been already reported on the [meep-discuss list](https://www.mail-archive.com/meep-discuss@ab-initio.mit.edu/msg05556.html), there is a bug which leads to field instabilities when combining 3 mirror-symmetry objects with periodic boundary conditions in 3d. 

This is demonstrated by the following simple example of a point source in vacuum:
```scm
(set-param! resolution 10)
(set! geometry-lattice (make lattice (size 1 1 1)))

(set! sources (list (src (make source (src (make gaussian-src (frequency 1) (fwidth 1))) (component Ez) (center 0 0 0)))))

(set! k-point (vector3 0 0 0))

(set! symmetries (list (make mirror-sym (direction X))
                       (make mirror-sym (direction Y))
                       (make mirror-sym (direction Z) (phase -1))))

(define print-field (lambda () (print "ez:, " (meep-time) ", " (magnitude (get-field-point Ez (vector3 0 0 0))) "\n")))

(run-until 500 (at-every 3.4 print-field))
```
For this script, the fields begin to blow up early on:
```
-----------
Initializing structure...
Working in 3D dimensions.
Computational cell is 1 x 1 x 1 with resolution 10
Halving computational cell along direction x
Halving computational cell along direction y
Halving computational cell along direction z
time for set_epsilon = 0.000983953 s
-----------
ez:, 3.4000000000000004, 0.11299371272955527
ez:, 6.800000000000001, 1.3292101237116931
ez:, 10.200000000000001, 2.402271717485686
ez:, 13.600000000000001, 4.484463313924006
ez:, 17.0, 3.8996648343918103
ez:, 20.400000000000002, 0.13634058216335865
ez:, 23.8, 4.629265409516416
ez:, 27.200000000000003, 4.998705102736996
ez:, 30.6, 1.2298673000561804
ez:, 34.0, 2.6981516161770243
ez:, 37.4, 4.377899104931889
ez:, 40.800000000000004, 3.5929413491696733
ez:, 44.2, 15.319503849169674
ez:, 47.6, 1194.3195038491704
ez:, 51.0, 2915498.319503849
ez:, 54.400000000000006, 559645525.6804962
ez:, 57.800000000000004, 4232176173909.6807
```
There are no instabilities if the Z mirror-symmetry object is removed:
```
-----------
Initializing structure...
Working in 3D dimensions.
Computational cell is 1 x 1 x 1 with resolution 10
Halving computational cell along direction x
Halving computational cell along direction y
time for set_epsilon = 0.00132418 s
-----------
ez:, 3.4000000000000004, 0.11299371272955483
ez:, 6.800000000000001, 1.3292101237116927
ez:, 10.200000000000001, 2.402271717485691
ez:, 13.600000000000001, 4.484463313924063
ez:, 17.0, 3.899664834392148
ez:, 20.400000000000002, 0.13634058216075873
ez:, 23.8, 4.629265409492891
ez:, 27.200000000000003, 4.998705102472007
ez:, 30.6, 1.2298672971468396
ez:, 34.0, 2.6981515888410907
ez:, 37.4, 4.3779221161355615
ez:, 40.800000000000004, 3.6004200025033697
ez:, 44.2, 0.1855690755252214
ez:, 47.6, 4.1906787491702815
ez:, 51.0, 5.302888125953832
ez:, 54.400000000000006, 1.6628091344931328
ez:, 57.800000000000004, 2.9025704016780374
```

### PR DESCRIPTION
**Pull request by oskooi**
Fixes #132.

**Root Cause**

The Bloch phase factor `eikna[d]` is computed using the **symmetry-reduced** grid volume (`gv`) size, while the periodic lattice translation vector (`ilattice_vector`) is computed using the **full** cell size (`user_volume`). This creates a mismatch between the phase applied per periodic translation and the actual translation distance.

**Why it only manifests with 2+ mirror symmetries + periodioc boundaries**

-  **single mirror** (e.g., `Mirror(Y)`) + **periodic in X**: only Y is halved, so `gv.num_direction(X)` equals the original full `user_volume.nx()`. The phase in X is correct. The Y direction typically has k_y = 0 with mirror symmetry, so `eikna[Y] = 1` regardless.
- **2+ mirrors** (e.g., `Mirror(X) + Mirror(Y)`) + **periodic**: Both X and Y are halved. If k_x ≠ 0, then `gv.num_direction(X)` ≈ `user_volume.nx() / 2` and `eikna[X]` is computed for the wrong period. The periodic translation moves by the full cell but applies the phase for half the cell. This produces incorrect field values at boundary conditions leading to expontentially growing errors (instability).
- **k = 0** (pure periodic, no Bloch phase): `eikna = exp(0) = 1` regardless of cell size, so the bug is latent and harmless.

Two new unit tests are added to compare the symmetry-exploiting simulation  against a reference simulation without symmetry, checking field values at multiple points and energies at regular intervals.

### CODE REVIEW

File: src/boundaries.cpp [hunk 1]
```diff
@@ -91,11 +91,15 @@ void fields::use_bloch(direction d, complex<double> kk) {
...
+  // Use user_volume (full cell) size, not gv (symmetry-reduced cell) size,
+  // because ilattice_vector and locate_point_in_user_volume translate by
+  // the full cell period. Using gv here would give the wrong Bloch phase
+  // when the cell is halved by mirror symmetries. (See issue #132.)
+  if (real(kk) * user_volume.num_direction(d) == 0.5 * a) // check b.z. edge exactly
+    eikna[d] = -exp(-imag(kk) * ((2 * pi / a) * user_volume.num_direction(d)));
   else {
     const complex<double> I = complex<double>(0.0, 1.0);
-    eikna[d] = exp(I * kk * ((2 * pi / a) * gv.num_direction(d)));
+    eikna[d] = exp(I * kk * ((2 * pi / a) * user_volume.num_direction(d)));
```
Comment (smartalecH) [inline src/boundaries.cpp:102]: Wow such a simple fix, for such a hairy bug all these years. Very cool!

Comment (stevengj) [reply src/boundaries.cpp:102]: The odd thing is that mirror symmetries (the halved dimensions) would only ever be in directions $i$ where $k_i = 0$, so it's not clear why this would have an effect?

Comment (stevengj) [reply src/boundaries.cpp:102]: However, it would affect you at $k_i = \pi / a$, which does preserve the mirror symmetry.

Comment (stevengj) [reply src/boundaries.cpp:102]: But then why would this help #132?

**Comment by stevengj** [comment]
Can you verify that this fixes #132?  It's not clear why it would help in a case where $k \ne 0$.

(But even if it fixed an unrelated bug that is nice too.)

**Comment by oskooi** [comment]
It seems the original test case in #132 with `k_point = (0, 0, 0)` actually runs fine (i.e. the fields do *not* blow up) even without this fix. The only test case in which the fields do blow up is `k_point = (0.5, 0.5, 0.5)` (i.e. at the edge of the Brillouin zone, also mentioned in #132) which is fixed by this PR.
